io_scene_godot/converters/material_node_tree/shaders.py

```diff
@@ -0,0 +1,536 @@
+"""Class represents fragment shader and vertex shader"""
+import re
+import collections
+import bpy
+import mathutils
+from .shader_functions import find_node_function_by_id
+from ...structures import Array, ValidationError
+
+
+class Variable:
+    """A variable in material shader scripts"""
+
+    def __init__(self, var_type, var_name):
+        self.type = var_type
+        self.name = var_name
+
+    def to_definition_str(self):
+        """Convert to a definition str"""
+        return self.type + ' ' + str(self)
+
+    def __str__(self):
+        """Convert to string"""
+        return self.name
+
+
+class Value:
+    """A constant value in material shader scripts"""
+
+    def __init__(self, type_str, data):
+        self.type = type_str
+        self.data = data
+
+    @classmethod
+    def create_from_blender_value(cls, blender_value):
+        """Creaate a Value() from a blender object"""
+        if isinstance(
+                blender_value, (bpy.types.bpy_prop_array, mathutils.Vector)):
+            tmp = list()
+            for val in blender_value:
+                tmp.append(val)
+
+            return Value("vec{}".format(len(tmp)), tuple(tmp))
+
+        if isinstance(blender_value, mathutils.Matrix):
+            # godot mat is column major order
+            mat = blender_value.transposed()
+            column_vec_list = list()
+            for vec in mat:
+                column_vec_list.append(cls.create_from_blender_value(vec))
+
+            return Value(
+                "mat{}".format(len(column_vec_list)),
+                tuple(column_vec_list)
+            )
+
+        return Value("float", blender_value)
+
+    def __str__(self):
+        """Convert to string"""
+        if self.type.startswith(('vec', 'mat')):
+            return "{}({})".format(
+                self.type,
+                ', '.join([str(x) for x in self.data])
+            )
+        return str(self.data)
+
+
+class FragmentOutClosure:
```

Comment:
The name Closure is a misnomer here. See https://en.wikipedia.org/wiki/Closure_(computer_programming)

To prevent confusion, maybe "FragmentBSDFContainer" would be better?

Reply:
yeah better than the current one